chain/indexer/manager.go

```diff
@@ -0,0 +1,204 @@
+}
+
+type Exporter interface {
+	ExportResult(ctx context.Context, strg model.Storage, m *ModelResult) error
+}
+
+// Manager manages the execution of an Indexer. It may be used to index TipSets both serially or in parallel.
+type Manager struct {
+	api      tasks.DataSource
+	storage  model.Storage
+	indexer  Indexer
+	exporter Exporter
+	window   time.Duration
+
+	// used for async tipset indexing
+	pool   *workerpool.WorkerPool
+	fatal  error
+	active int64
+}
+
+type ManagerOpt func(i *Manager)
+
+// WithWorkerPool overrides the Manager's default pool (1) with the provided pool.
+// The size of the pool controls the number of TipSets that can be indexed in parallel.
+func WithWorkerPool(pool *workerpool.WorkerPool) ManagerOpt {
+	return func(m *Manager) {
+		m.pool = pool
+	}
+}
+
+// WithWindow overrides the Manager's default (0) window with the provided window.
+// The value of the window controls a timeout after which the Manager aborts processing the tipset, marking any incomplete
+// tasks as SKIPPED.
+func WithWindow(w time.Duration) ManagerOpt {
+	return func(m *Manager) {
+		m.window = w
+	}
+}
+
+// WithExporter overrides the Manager's default Exporter with the provided Exporter.
+// An Exporter is used to export the results of the Manager's Indexer.
+func WithExporter(e Exporter) ManagerOpt {
+	return func(m *Manager) {
+		m.exporter = e
+	}
+}
+
+// WithIndexer overrides the Manager's default Indexer with the provided Indexer.
+// An Indexer is used to collect state from a tipset.
+func WithIndexer(i Indexer) ManagerOpt {
+	return func(m *Manager) {
+		m.indexer = i
+	}
+}
+
+// NewManager returns a default Manager. Any provided ManagerOpt's will override Manager's default values.
+func NewManager(api tasks.DataSource, strg model.Storage, name string, tasks []string, opts ...ManagerOpt) (*Manager, error) {
+	indexer, err := NewTipSetIndexer(api, name, tasks)
+	if err != nil {
+		return nil, err
+	}
+
+	im := &Manager{
+		api:      api,
+		storage:  strg,
+		indexer:  indexer,
+		exporter: NewModelExporter(),
+		window:   0,
+		pool:     workerpool.New(1),
+	}
+	for _, opt := range opts {
+		opt(im)
+	}
+	return im, nil
+}
+
+// TipSetAsync enqueues `ts` into the Manager's worker pool for processing. An error is returned if any Indexer in
+// the pool encounters a fatal error, else the method returns immediately.
+func (i *Manager) TipSetAsync(ctx context.Context, ts *types.TipSet) error {
+	if i.fatal != nil {
+		defer i.pool.Stop()
+		return i.fatal
+	}
+
+	stats.Record(ctx, metrics.IndexManagerActiveWorkers.M(i.active))
+	stats.Record(ctx, metrics.IndexManagerWaitingWorkers.M(int64(i.pool.WaitingQueueSize())))
+	if i.pool.WaitingQueueSize() > i.pool.Size() {
+		log.Warnw("queuing worker in Manager pool", "waiting", i.pool.WaitingQueueSize())
+	}
+	log.Infow("submitting tipset for async indexing", "height", ts.Height(), "active", i.active)
+
+	ctx, span := otel.Tracer("").Start(ctx, "Manager.TipSetAsync")
+	i.pool.Submit(func() {
```

Comment:
This pool used to live in the [watcher](https://github.com/filecoin-project/lily/blob/fe9a838028cb447ae54e8af0b5349d57c7912427/chain/watcher.go#L134), its now managed here instead.